src/openzeppelin/access/ownable.cairo

```diff
@@ -80,6 +104,73 @@ namespace Ownable:
         return ()
     end
 
+    func propose_owner{
+            syscall_ptr : felt*,
+            pedersen_ptr : HashBuiltin*,
+            range_check_ptr
+        }(proposed_owner: felt):
+        with_attr error_message("Ownable: proposed owner cannot be the zero address"):
+            assert_not_zero(proposed_owner)
+        end
+
+        _propose_owner(proposed_owner)
+        return()
+    end
+
+    func accept_ownership{
+            syscall_ptr : felt*,
+            pedersen_ptr : HashBuiltin*,
+            range_check_ptr
+        }():
+        let (proposed_owner) = Ownable.proposed_owner()
+        let (caller) = get_caller_address()
+        # caller cannot be zero address to avoid overwriting owner when proposed_owner is not set
+        with_attr error_message("Ownable: caller is the zero address"):
+            assert_not_zero(caller)
+        end
+
+        with_attr error_message("Ownable: caller is not the proposed owner"):
+            assert caller = proposed_owner
+        end
+
+        _transfer_ownership(proposed_owner)
+        return()
+    end
+
+    func cancel_ownership_proposal{
+            syscall_ptr : felt*,
+            pedersen_ptr : HashBuiltin*,
+            range_check_ptr
+        }():
+        alloc_locals
+        let (local proposed_owner) = Ownable.proposed_owner()
+
+        with_attr error_message("Ownable: proposed owner is already the address zero, no need to cancel"):
+            assert_not_zero(proposed_owner)
+        end
+
+        let (local current_owner) = Ownable.owner()
+        let (local caller) = get_caller_address()
+
+        tempvar syscall_ptr = syscall_ptr
+        tempvar pedersen_ptr = pedersen_ptr
+        tempvar range_check_ptr = range_check_ptr
+
+        # Can only be called by current owner or proposed owner
+        with_attr error_message("Ownable: caller is neither the current owner nor the proposed owner"):
+            if caller != proposed_owner:
+                if caller != current_owner:
+                    assert 1 = 0
+                end
+            end
+        end
+
+        Ownable_proposed_owner.write(0)
```

Comment:
> @ctrlc03 good work on adding in the tests! If you don't mind taking a look, I left some questions and suggestions.
> 
> Also, I think this two-step feature should be an extension as opposed to living in `Ownable.cairo`. @JulissaDantes @frangio thoughts? Maybe `OwnableTwoStepTransfer.cairo` or `OwnableTwoStep.cairo`?
> 
> If this is going to live in Ownable, we should reset the proposed owner within `_transfer_ownership`. If it's going to be an extension, we'll have to reset the the proposed owner upon transfers another way—otherwise, the proposed owner could still accept the ownership transfer after it's already been transferred (via `transferOwnership`).

I've added the reset of the `Ownable_proposed_owner` inside `transferOwnership` for now. Happy to move this PR to a separate library if that works better though.

Reply:
@andrew-fleming After adding the `_reset_proposed_owner` inside the `transfer_ownership`, I have a revoked reference on the `previous_owner`. I find it weird that adding `alloc_locals` without setting the variable to `local` works. Any thought on that? Or did I miss some feature changes in Cairo or have I always misunderstood this? 

Btw, all test cases have been added. Hopefully now this is good and it's only missing the decision of moving it to a separate library or not.